source/android/adaptivecards/src/main/java/com/microsoft/adaptivecards/renderer/registration/CardRendererRegistration.java

```diff
@@ -0,0 +1,132 @@
+        m_typeToRendererMap.put(CardElementType.ColumnSet.swigValue(), ColumnSetRenderer.getInstance());
+        m_typeToRendererMap.put(CardElementType.Container.swigValue(), ContainerRenderer.getInstance());
+        m_typeToRendererMap.put(CardElementType.FactSet.swigValue(), FactSetRenderer.getInstance());
+        m_typeToRendererMap.put(CardElementType.Image.swigValue(), ImageRenderer.getInstance());
+        m_typeToRendererMap.put(CardElementType.ImageSet.swigValue(), ImageSetRenderer.getInstance());
+        m_typeToRendererMap.put(CardElementType.TextBlock.swigValue(), TextBlockRenderer.getInstance());
+
+        // Register Input Renderers
+        m_typeToRendererMap.put(CardElementType.TextInput.swigValue(), TextInputRenderer.getInstance());
+        m_typeToRendererMap.put(CardElementType.NumberInput.swigValue(), NumberInputRenderer.getInstance());
+        m_typeToRendererMap.put(CardElementType.DateInput.swigValue(), DateInputRenderer.getInstance());
+        m_typeToRendererMap.put(CardElementType.TimeInput.swigValue(), TimeInputRenderer.getInstance());
+        m_typeToRendererMap.put(CardElementType.ToggleInput.swigValue(), ToggleInputRenderer.getInstance());
+        m_typeToRendererMap.put(CardElementType.ChoiceSetInput.swigValue(), ChoiceSetInputRenderer.getInstance());
+    }
+
+    public static CardRendererRegistration getInstance()
+    {
+        if (s_instance == null)
+        {
+            s_instance = new CardRendererRegistration();
+        }
+
+        return s_instance;
+    }
+
+    public void registerRenderer(CardElementType cardElementType, IBaseCardElementRenderer renderer)
+    {
+        if (cardElementType != null || cardElementType == CardElementType.Unsupported)
+        {
+            throw new IllegalArgumentException("cardElementType is null or unsupported");
+        }
+        if (renderer == null)
+        {
+            throw new IllegalArgumentException("renderer is null");
+        }
+
+        m_typeToRendererMap.put(cardElementType.swigValue(), renderer);
+    }
+
+    public IBaseCardElementRenderer getRenderer(CardElementType cardElementType)
+    {
+        return m_typeToRendererMap.get(cardElementType.swigValue());
+    }
+
+    public View render(
+            Context context,
+            FragmentManager fragmentManager,
+            ViewGroup viewGroup,
+            Object tag,
+            BaseCardElementVector baseCardElementList,
+            Vector<IInputHandler> inputActionHandlerList,
+            HostConfig hostConfig)
+    {
+        long size;
+        if (baseCardElementList == null || (size = baseCardElementList.size()) <= 0)
+        {
+            return viewGroup;
+        }
+
+        LinearLayout layout = new LinearLayout(context);
+        layout.setTag(tag);
+        layout.setLayoutParams(new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT));
```

Comment:
it's one line of code that can't really be simplified. My guideline to split into helper function when there is multi-line and mostly with some operational aspect that's repeated.

---
In reply to: [126744523](https://github.com/Microsoft/AdaptiveCards/pull/461#discussion_r126744523) [](ancestors = 126744523)